src/main/java/org/jenkinsci/plugins/workflow/job/console/NewNodeConsoleNote.java

```diff
@@ -0,0 +1,125 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.job.console;
+
+import hudson.Extension;
+import hudson.MarkupText;
+import hudson.Util;
+import hudson.console.ConsoleAnnotationDescriptor;
+import hudson.console.ConsoleAnnotator;
+import hudson.console.ConsoleNote;
+import hudson.model.TaskListener;
+import java.io.IOException;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.jenkinsci.plugins.workflow.actions.LabelAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.support.actions.AnnotatedLogAction;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Console line with note printed when a new {@link FlowNode} is added to the graph.
+ * Defines the {@code pipeline-new-node} CSS class and several attributes which may be used to control subsequent behavior:
+ * <ul>
+ * <li>{@code nodeId} for {@link FlowNode#getId}
+ * <li>{@code parentIds} for {@link FlowNode#getParents}
+ * <li>{@code startId} {@link FlowNode#getId} for {@link BlockStartNode}, else {@link BlockEndNode#getStartNode}, else absent
+ * <li>{@code label} for {@link LabelAction} if present
+ * </ul>
+ * @see AnnotatedLogAction#annotateHtml
+ */
+@Restricted(NoExternalUse.class)
+public class NewNodeConsoleNote extends ConsoleNote<WorkflowRun> {
+
+    /**
+     * Prefix used in metadata lines.
+     */
+    private static final String CONSOLE_NOTE_PREFIX = "[Pipeline] ";
+
+    public static void print(FlowNode node, TaskListener listener) {
+        try {
+            listener.annotate(new NewNodeConsoleNote(node));
+        } catch (IOException x) {
+            // never mind
+        }
+        listener.getLogger().println(CONSOLE_NOTE_PREFIX + node.getDisplayFunctionName());
+    }
+
+    private final @Nonnull String id;
+    private final @Nonnull String[] parents;
```

Comment:
Perhaps just replace with `@CheckForNull String enclosing` (mutually exclusive with `start`) so that `script.js` does not need to reimplement `FlowScanningUtils.fetchEnclosingBlocks`.